### PR TITLE
MWPW-134799: MEP On/Off

### PR DESCRIPTION
**Pull request by 3ch023**
## MEP On/Off
This PR introduces a client-side on/off feature for MEP manifests.
[Related Discussion](https://github.com/orgs/adobecom/discussions/1384#discussioncomment-7634206)
More details below.

## Resolves: 
- [MWPW-134799](https://jira.corp.adobe.com/browse/MWPW-134799)
- [MWPW-137499](https://jira.corp.adobe.com/browse/MWPW-137499)

## Test URLs:
- Before: https://main--milo--adobecom.hlx.page/products/special-offers?martech=off
- After: https://onoff--milo--adobecom.hlx.page/products/special-offers?martech=off

### [Promotions.xlsx](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/Doc.aspx?sourcedoc=%7BCF2BC44F-DE66-4C08-A1CA-98CAD55ECC61%7D&file=promotions.xlsx&action=default&mobileredirect=true)
_Promotions.xlsx_ - a separate global metadata file, configured in `.helix/config.xlsx`

image

- located in root (milo folder)
- holds connection `folder/page`-`manifest` and on/off date times.
- serves metadata from the 'helix-default' tab
- has the 'schedule' tab to configure on/off dates
- The 'helix-default' has a hidden column called `'scheduled'`. It is auto-generated from the 'schedule' tab and only visible when you preview _promotions.xlsx_.

image

Why a separate metadata file?
- avoid conflicts & risks when GWP is modifying the main metadata file to configure promotions
- easier overview & troubleshooting for promo issues
- potential automation data file

## Schedule Tab
image

- no row/column can be deleted -> that would break generation functions in the 'concut' tab
- datepicker & formatting for date fields
- time formating for time fields
- active/inactive controls if the schedule line is applied to metadata. GWP can set a line to inactive if they want to re-enable it later.

## Priority
_Promotions.xlsx_ is higher priority than metdata.xlsx 
_Promotions.xlsx_ only allows 2 fields to be configured, 'personalization' and 'events'. Its protected, so no other column from global metadata can be overwritten. On the contrary, global metadata should not be able to overwrite the 'personalization' and 'events' data in promotions.xlsx.
Priority is achieved due to order in .helix/config.xlsx

## 'manifestnames' and 'schedule' meta tags
- 'manifestnames' contains manifests attached to this particular page (similar to 'personalization' field, but instead of path GWP can use a so-called manifest nickname, from the schedule tab)
- 'schedule' contains all active promo schedule lines
- 'schedule' line format: `manifest name` | `start datetime (GMT)` | `end datetime (GMT)` | `manifest path`
[Sample page](https://onoff--milo--adobecom.hlx.page/products/special-offers?martech=off)

```
<meta name="manifestnames" content="pre-black-friday-global,black-friday-global,cyber-monday">
<meta name="schedule" content="pre-black-friday-global | 2023-11-01T00:00:00 | 2023-12-15T00:00:00 | https://main--milo--adobecom.hlx.page/promos/2023/black-friday/pre-black-friday/manifest-global.json, black-friday-global | 2023-12-15T00:00:00 | 2023-12-31T00:00:00 | https://main--milo--adobecom.hlx.page/promos/2023/black-friday/black-friday/manifest-global.json">
```

## MEP behavior

- MEP will read 'schedule' metadata
- MEP will sort out of 'schedule' promo manifests attached to this page
- For each manifest it will decide if its enabled (scheduled for current time) or disabled (scheduled outside of current time or the schedule is missing at all)
- MEP will handle the manifests from 'personalization' and the ones from 'promo' in the same way. The 'personalization' manifests are lower priority - **to be discussed.**
- if 'bts' event is found, the start and end datetime are converted to a local browser time, then compared with the current time. If it's on - manifest is enabled, if not - disabled.
- disabled manifest is **NOT preloaded** and **NOT fetched**. This is true for both hlx.page and hlx.live.
- Preview for promo manifests work the same as for usual personalization.

## Preview
'Scheduled - inactive/active' shows if the current browser time is within the manifest's on/off period.
On and Off date time is shown in the local timezone, NOT in GMT.
When there is a manifest overwrite in the URL, MEP will disregard the schedule. This is needed for QA-ing inactive changes.
image

### CODE REVIEW

File: libs/features/personalization/manifest-utils.js [hunk 1]
```diff
@@ -19,15 +20,46 @@ export const normalizePath = (p) => {
...
+    return { name, start: GMTStringToLocalDate(start), end: GMTStringToLocalDate(end) };
+  }) : [];
+};
+
+const getEventInfo = (manifestPath, events) => {
+  let disabled = false;
+  let event = null;
+  const scheduled = manifestPath.match(REGEX_PROMO);
+  if (scheduled) {
+    event = events.find((e) => e.name === scheduled[1]);
```
Comment (npeltier) [inline libs/features/personalization/manifest-utils.js:37]: what is 1? could we use a talkative constant (PROMO_NAME_GRP?)

File: libs/features/personalization/personalization.js [hunk 2]
```diff
@@ -559,13 +561,28 @@ export async function applyPers(manifests) {
   const cleanedManifests = cleanManifestList(manifests);
 
   let results = [];
+  const experiments = [];
   for (const manifest of cleanedManifests) {
-    results.push(await runPersonalization(manifest, config));
+    const noOverride = !config.mep?.override;
```
Comment (npeltier) [inline libs/features/personalization/personalization.js:566]: wouldn't it be clearer to have an override flag, and then !override test?

File: libs/features/personalization/manifest-utils.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
-import { loadLink } from '../../utils/utils.js';
+import { loadLink, getMetadata } from '../../utils/utils.js';
 
+const REGEX_PROMO = /\/promos\/(\w+)\//;
```
Comment (npeltier) [inline libs/features/personalization/manifest-utils.js:3]: is \w+ enough for first level folder? characters like '-' or '_' are allowed, no?

File: libs/features/personalization/manifest-utils.js [hunk 4]
```diff
@@ -19,15 +20,46 @@ export const normalizePath = (p) => {
...
+  let disabled = false;
+  let event = null;
+  const scheduled = manifestPath.match(REGEX_PROMO);
+  if (scheduled) {
+    event = events.find((e) => e.name === scheduled[1]);
+    if (!event) {
+      disabled = true;
+    } else {
+      const currentDate = new Date();
+      if (event.start && event.end
```
Comment (npeltier) [inline libs/features/personalization/manifest-utils.js:42]: don't we want to manage "evergreen" promotion that are endless?

Comment (3ch023) [reply libs/features/personalization/manifest-utils.js:42]: for now, i don't recall such use-case.
can you give more details?
how will it work?

Comment (npeltier) [reply libs/features/personalization/manifest-utils.js:42]: never mind those use case are not meant to leverage mep anyway.

File: libs/features/personalization/personalization.js (hunk 2 above)
Comment (3ch023) [reply libs/features/personalization/personalization.js]: fixed

**Comment by 3ch023** [comment]
for now rebased on `mwpw-138750-analytics2` as agreed with @vivgoodrich, 
once [pull/1531](https://github.com/adobecom/milo/pull/1531) is merged, this PR will be rebased to the main and ready to merge.

File: libs/features/personalization/manifest-utils.js (hunk 4 above)
Comment (honstar) [reply libs/features/personalization/manifest-utils.js]: > don't we want to manage "evergreen" promotion that are endless?

Evergreen sounds like regular content then, no? Should not be tied to promotion anymore in that case.

**Review by vgoodric — APPROVED** [review]

**Review by chrischrischris — APPROVED** [review]

**Pull request merged**